File: problems/0189-rotate-array/solution.go
func rotate(nums []int, k int) {
    // naive solution
    // []
    // [1,2,3,4,5,6,7]

    // [5], append nums[len(nums) - k + i] for 0 <= i < k
    // [5], append nums[i] for 0 <= i < n-k
    // [1,2,3,4,5,6,7]
    result := []int{}
    n := len(nums)
    if k > n {
        k = k % n
    }

    for i := 0; i < k; i++ {
        result = append(result, nums[n-k+i])
    }

    for i := 0; i < n-k; i++ {
        result = append(result, nums[i])
    }
    copy(nums, result)
    
}

File: problems/0189-rotate-array/solution.cpp
class Solution {
public:
    void rotate(vector<int>& nums, int k) {
        // append nums[n - k + i] for 0 <= i < k to new array
        // append nums[i] for 0 <= i < n-k to new array
        // copy the new array over into nums
        // O(n) space complexity
        // O(n) time complexity
        int n = nums.size();
        k = k % n;

        vector<int> tmp;

       
        for (int i = 0; i < k; i++) {
            tmp.push_back(nums[n-k+i]);
        }
        for (int i = 0; i < n-k; i++) {
            tmp.push_back(nums[i]);
        }
        nums = tmp;
    }
};
